### PR TITLE
Execution with heaptrack

### PR DESCRIPTION
**Pull request by partouf**
Should allow us to offer:
* Changing environment variables for user executable executions
* Enable Heaptrack for Linux (graph via speedscope, summary, details)
  - Heaptrack runs via LD_PRELOAD and pipes allocation info to a file
* Enable future adding of similar tooling (probably not at the same time, but it depends on how the tool works)
* Enable future adding of process statistics under Windows

### CODE REVIEW

**Comment by partouf** [comment]
Screenshot from 2023-10-24 03-52-46

The stderr from the execution contains a summary, the popup offers a link to speedscope

**Comment by partouf** [comment]
Also offers `-ftime-trace` in speedscope instead of perfetto as requested in https://github.com/compiler-explorer/compiler-explorer/pull/4967#issuecomment-1525952200

File: static/panes/compiler.ts [hunk 1]
```diff
@@ -1768,14 +1768,46 @@ export class Compiler extends MonacoPane<monaco.editor.IStandaloneCodeEditor, Co
...
+    offerViewInSpeedscope(artifact: Artifact): void {
+        this.alertSystem.notify(
+            'Click ' +
+                '<a target="_blank" id="download_link" style="cursor:pointer;" click="javascript:;">here</a>' +
+                ' to view ' +
+                artifact.title +
+                ' in Speedscope',
+            {
+                group: 'emulation',
+                collapseSimilar: true,
```
Comment (partouf) [inline static/panes/compiler.ts:1790]: should not collapse (and groupname should not be emulation)

File: lib/heaptrack/heaptrack-wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            file.close(err => {
+                if (err) logger.error('Error while closing heaptrack log: ', err);
+                resolve(true);
+            });
+        });
+
+        await new Promise(resolve => fs.close(fd, () => resolve(true)));
+    }
+
+    public async exec(filepath: string, args: string[], execOptions: ExecutionOptions): Promise<UnprocessedExecResult> {
```
Comment (partouf) [inline lib/heaptrack/heaptrack-wrapper.ts:78]: should split up this function more

File: lib/heaptrack/heaptrack-wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    private add_to_env(execOptions: ExecutionOptions) {
+        if (!execOptions.env) execOptions.env = {};
+
+        if (execOptions.env.LD_PRELOAD) {
+            execOptions.env.LD_PRELOAD = this.preload + ':' + execOptions.env.LD_PRELOAD;
+        } else {
+            execOptions.env.LD_PRELOAD = this.preload;
+        }
+
+        execOptions.env.DUMP_HEAPTRACK_OUTPUT = '/app/heaptrack_fifo';  // todo: will not work for local users without nsjail, is that a problem?
```
Comment (partouf) [inline lib/heaptrack/heaptrack-wrapper.ts:50]: https://github.com/compiler-explorer/compiler-explorer/pull/5533/files#diff-23200c2e7936d6738c5056e91063698d8d9db45b30fa99be68e22eed138ee8eaR219 offers code to identify the sandboxtype, should become its own PR that we can merge in here and use for this

File: lib/heaptrack/heaptrack-wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    private sandboxFunc: TypicalExecutionFunc;
+    private execFunc: TypicalExecutionFunc;
+
+    public static FlamegraphFilename = 'heaptrack.flamegraph.txt';
+
+    constructor(dirPath: string, sandboxFunc: TypicalExecutionFunc, execFunc: TypicalExecutionFunc) {
+        this.dirPath = dirPath;
+        this.pipe = path.join(this.dirPath, 'heaptrack_fifo');
+        this.raw_output = path.join(this.dirPath, 'heaptrack.raw');
+        this.heaptrackPath = '/opt/compiler-explorer/heaptrack/1.3.0'; // todo: needs to be a property
```
Comment (partouf) [inline lib/heaptrack/heaptrack-wrapper.ts:29]: use `prop()`

File: lib/base-compiler.ts [hunk 5]
```diff
@@ -1960,6 +1998,9 @@ export class BaseCompiler implements ICompiler {
         }
 
         const result = await this.doExecution(key, executeParameters, bypassCache);
+        if (executeParameters.analysis?.includes('heaptrack')) {
```
Comment (partouf) [inline lib/base-compiler.ts:2001]: put text in some const or enum we can point to instead of passing text

File: lib/base-compiler.ts [hunk 6]
```diff
@@ -868,11 +884,11 @@ export class BaseCompiler implements ICompiler {
             .filter(link => link) as string[];
     }
 
-    getSharedLibraryPaths(libraries: CompileChildLibraries[]) {
+    getSharedLibraryPaths(libraries: CompileChildLibraries[]): string[] {
         return libraries
             .map(selectedLib => {
                 const foundVersion = this.findLibVersion(selectedLib);
-                if (!foundVersion) return false;
+                if (!foundVersion) return [];
```
Comment (partouf) [inline lib/base-compiler.ts:891]: this bugfix and the introduction of types can be a separate PR

**Comment by jeremy-rifkin** [comment]
Looks awesome

**Comment by partouf** [comment]
In the CI there's an error regarding not being able to find mkfifo - might be because this is an old package and it's not configured to be modular? I'm curious why it works locally for me... then again I do still have node 18 and CI tries it with node 20

**Comment by partouf** [comment]
I just realized... mkfifo does of course not work on windows...

Probably will just need to do a exec call to create a pipe instead of using the module

**Comment by partouf** [comment]
Progress so far:

Screenshot from 2023-10-27 22-50-59

**Comment by partouf** [comment]
Not added to Compiler pane, but think this is good enough to release

**Comment by partouf** [comment]
Think we can also make ENV vars possible to add to Favorites in this screen, but can be left for later

**Comment by partouf** [comment]
Also
Fixes #3626
Fixes #3046

**Comment by partouf** [comment]
Screenshot from 2023-11-01 16-30-29

**Comment by partouf** [comment]
(Accidentally) tested this on staging, but am getting:

```
ERROR: ld.so: object '/opt/compiler-explorer/heaptrack/v1.3.0/lib/libheaptrack_preload.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
[F][2023-11-04T21:02:17+0000][1] bool subproc::runChild(nsjconf_t*, int, int, int, int)():449 Launching child process failed
```

Confused...

**Comment by partouf** [comment]
> Confused...

OH the path is wrong
should be /opt/compiler-explorer/heaptrack-v1.3.0

**Review by mattgodbolt — APPROVED** [review]
I didn't read it all. it's a big PR and touches a bunch of things...I trust you to test it :)

File: static/panes/executor.ts [hunk 7]
```diff
@@ -678,6 +680,48 @@ export class Executor extends Pane<ExecutorState> {
...
+                artifact.title +
+                ' in Speedscope',
+            {
+                group: artifact.type,
+                collapseSimilar: false,
+                dismissTime: 10000,
+                onBeforeShow: function (elem) {
+                    elem.find('#download_link').on('click', () => {
+                        const tmstr = Date.now();
+                        const live_url = 'https://static.ce-cdn.net/speedscope/index.html';
```
[the remark this answers is not in the document]
Comment (partouf) [reply static/panes/executor.ts:711]: The official version that's online only allows local or remote file urls, no passing of files to the website directly (I think). There's an upload thing, but I don't think we can access that. I could also send a PR, but the repo's readme says we should just fork.

File: lib/runtime-tools/heaptrack-wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        this.interpreter = path.join(this.heaptrackPath, 'libexec/heaptrack_interpret');
+        this.printer = path.join(this.heaptrackPath, 'bin/heaptrack_print');
+    }
+
+    public static isSupported(compiler: CompilationEnvironment) {
+        return process.platform !== 'win32' && compiler.ceProps('heaptrackPath', '') !== '';
+    }
+
+    private async mkfifo(path: string, rights: number) {
+        await executeDirect(this.mkfifoPath, ['-m', rights.toString(8), path], {});
```
[the remark this answers is not in the document]
Comment (partouf) [reply lib/runtime-tools/heaptrack-wrapper.ts:83]: Right, I could have also forked https://www.npmjs.com/package/mkfifo (it basically just does the linux call https://linux.die.net/man/3/mkfifo) and make a dummy windows implementation, but that's... well... time consuming...

I don't know about deadlocks, the only issue I encountered during testing was incompleteness and not being able to delete the file, but that was fixed by reordering some of the close/free/release calls.

[the remark this answers is not in the document]
Comment (mattgodbolt) [reply lib/runtime-tools/heaptrack-wrapper.ts:83]: We'll see how this goes. The functionality is worth getting out in the field and debugging afterwards :)

[the remark this answers is not in the document]
Comment (mattgodbolt) [reply lib/runtime-tools/heaptrack-wrapper.ts:83]: (if even needed!)
